### PR TITLE
add Nanopore support: Flye and Medaka

### PR DESCRIPTION
**Pull request by Emkago**
# Description
Added tools for genome assembly from Nanopore data:
- added 2 modules: Flye and Medaka
- updated config files (base and hopper)
- updated bacterial_base.nf

## Primary function of PR
- [ ] Hotfix
- [ ] Minor functionality improvement
- [x] Major functionality improvement / New type of analysis
- [ ] Backward-breaking functionality

Any suggestions how --genome-size in Flye could be defined as a parameter for each species?

# Testing
Has not been tested yet.

### CODE REVIEW

**Review by sylvinite — APPROVED** [review]
Wow nicely done! Now we're up to 3 assemblers! 💪 

It is worthwhile in the future to expand the README to make instructions for Nanopore vs others a bit more explicit.

File: README.md [hunk 1]
```diff
@@ -54,6 +54,10 @@ Source: `configs/nextflow.base.config`
 * Edit the `krakenDb`, `workDir` and `outdir` parameters in `configs/nextflow.base.config`
 * Edit the `runOptions` in `configs/nextflow.base.config` in order to mount directories to your run
 
+When analysing Nanopore data:
+* Edit the `ext.args` for Flye: specify genome size for the organism of interest with flag `--genome-size`.
+* Edit the `ext.args` for Medaka: specify the model with flag `-m`. Currently it is set to `r941_min_sup_g507`, but one should always set it based on how the data was produced. More about choosing the right model can be found [here](https://github.com/nanoporetech/medaka#models).
```
Comment (ryanjameskennedy) [inline README.md:59]: Might be nice to do a table similar to that of Usage argumentsin line 97 for all the possible arguments. I'll make an issue and do it when I have spare time!

File: nextflow-modules/modules/flye/main.nf [hunk 2]
```diff
@@ -0,0 +1,42 @@
+process flye {
+  tag "${sampleName}"
+  scratch params.scratch
+
+  input:
+    tuple val(sampleName), path(reads), val(platform), val(mode)
```
Comment (ryanjameskennedy) [inline nextflow-modules/modules/flye/main.nf:6]: mode needs to be removed from this input channel:)

**Review by ryanjameskennedy — CHANGES_REQUESTED** [review]
Everything looks great, the mode argument from flye just needs to be moved to the config:)

File: nextflow-modules/modules/flye/main.nf [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    path "*versions.yml"                                       , emit: versions
+
+  when:
+    task.ext.when && platform == "nanopore"
+
+  script:
+    def args = task.ext.args ?: ''
+    outputDir = params.publishDir ? params.publishDir : 'flye'
+    """
+    flye ${mode} ${reads} ${args} -o ${outputDir}
```
[the remark this answers is not in the document]
Comment (Emkago) [reply nextflow-modules/modules/flye/main.nf]: Changed ${mode} to `--nano-hq` as I it seems this is the mode that suits newly produced Nanopore data. If there is a need for a different mode, a possibility to choose a mode can be added.

Info from [Flye manual](https://github.com/fenderglass/Flye/blob/flye/docs/USAGE.md#oxford-nanopore):

The default mode for regular ONT data is --nano-raw. It works well for a good range of datasets, from old R7 pores to the most recent R9.x and R10.x. The expected error rate is 10-15%.

For the most recent ONT data basecalled with Guppy5+ SUP use the new --nano-hq mode. Expected error rate is <5%.

For Q20 data, use a combination of --nano-hq and --read-error 0.03.

If you have error-corrected ONT reads (with methods such as Canu), use --nano-corr.

[the remark this answers is not in the document]
Comment (ryanjameskennedy) [reply nextflow-modules/modules/flye/main.nf]: In order for users to be able to change this easily it needs to be added to the configs under flye's configurations like this:
```
ext.args = "--nano-hq --threads 40 --genome-size 2.8m --asm-coverage 50"
```

**Review by ryanjameskennedy — CHANGES_REQUESTED** [review]
Just need to move the `--nano-hq` to the config external arguments.

File: nextflow-modules/modules/flye/main.nf (hunk 3 above)
[the remark this answers is not in the document]
Comment (Emkago) [reply nextflow-modules/modules/flye/main.nf]: Done.

File: nextflow-modules/modules/flye/main.nf [hunk 4]
```diff
@@ -16,7 +16,7 @@ process flye {
     def args = task.ext.args ?: ''
     outputDir = params.publishDir ? params.publishDir : 'flye'
     """
-    flye ${mode} ${reads} ${args} -o ${outputDir}
+    flye --nano-hq ${reads} --out-dir ${outputDir} ${args}
```
[the remark this answers is not in the document]
Comment (Emkago) [reply nextflow-modules/modules/flye/main.nf]: Done.

**Review by ryanjameskennedy — APPROVED** [review]
LGTM!

**Comment by Emkago** [comment]
> LGTM! After a quick stub-run should be good to merge!

I am on it. A run starts, but Flye and Medaka are not executed. Have to investigate why.

File: workflows/bacterial_base.nf [hunk 5]
```diff
@@ -2,6 +2,7 @@
 
 nextflow.enable.dsl=2
 
+<<<<<<< HEAD
```
Comment (ryanjameskennedy) [inline workflows/bacterial_base.nf:5]: You need to resolve this merge conflict before commiting/pushing.

**Review by ryanjameskennedy — CHANGES_REQUESTED** [review]
Just want to block the merging since this was approved ages ago and now does not meet the requirements.

File: nextflow-modules/modules/medaka/main.nf [hunk 6]
```diff
@@ -4,14 +4,15 @@ process medaka {
 
   input:
     tuple val(sampleName), path(reads), val(platform) 
-    path assembly
-
+    // path assembly
```
Comment (ryanjameskennedy) [inline nextflow-modules/modules/medaka/main.nf:7]: You can remove these if they don't work

File: nextflow-modules/modules/flye/main.nf [hunk 7]
```diff
@@ -10,13 +10,15 @@ process flye {
     path "*versions.yml"                                       , emit: versions
 
   when:
-    task.ext.when && platform == "nanopore"
+    // task.ext.when && platform == "nanopore" // process was not launched
```
Comment (ryanjameskennedy) [inline nextflow-modules/modules/flye/main.nf:13]: Remove if it doesn't work

**Review by ryanjameskennedy — CHANGES_REQUESTED** [review]
Just a few coments. I think we should add a test csv for nanopore as well as test data. Perhaps we do that in another PR though.

**Comment by Emkago** [comment]
After a few more changes in modules and base config assembly of Nanopore data works. Medaka runs two times as suggested by Amaya to get out a better assembly. I have tested it with S. aureus long-read data produced locally in our lab.

**Review by sylvinite — APPROVED** [review]

**Pull request merged**